apps/common/lib/lexical/ast/analysis/analyzer.ex

```diff
@@ -26,26 +26,211 @@ defmodule Lexical.Ast.Analysis.Analyzer do
+          |> MapSet.union(functions)
+
+        [only: :macros] ->
+          current_imports
+          |> remove_module_imports(import_module)
+          |> MapSet.union(macros)
+
+        [only: fa_list] ->
+          fa_mapset = function_and_arity_to_mfa(import_module, fa_list)
+
+          current_imports
+          |> remove_module_imports(import_module)
+          |> MapSet.union(fa_mapset)
+
+        [except: fa_list] ->
+          # This one is a little tricky. Imports using except have two cases.
+          # In the first case, if the module hasn't been previously imported, we
+          # collect all the functions in the current module and remove the ones in the
+          # except clause.
+          # If the module has been previously imported, we just remove the functions from
+          # the except clause from those that have been previously imported.
+          # See: https://hexdocs.pm/elixir/1.13.0/Kernel.SpecialForms.html#import/2-selector
+
+          fa_mapset = function_and_arity_to_mfa(import_module, fa_list)
+
+          current_imports =
+            if already_imported?(current_imports, import_module) do
+              MapSet.difference(current_imports, fa_mapset)
+            else
+              current_imports
+              |> MapSet.union(functions)
+              |> MapSet.union(macros)
+            end
+
+          MapSet.difference(current_imports, fa_mapset)
+      end
+    end
+
+    defp expand_selector(selectors) do
+      selectors =
+        Enum.reduce(selectors, [], fn
+          {{:__block__, _, [type]}, {:__block__, _, [selector]}}, acc
+          when type in [:only, :except] ->
+            expanded =
+              case selector do
+                :functions ->
+                  :functions
+
+                :macros ->
+                  :macros
+
+                keyword when is_list(keyword) ->
+                  Enum.map(keyword, fn
+                    {{:__block__, _, [function_name]}, {:__block__, _, [arity]}} ->
+                      {function_name, arity}
+                  end)
+              end
+
+            [{type, expanded} | acc]
+
+          _, acc ->
+            acc
+        end)
+
+      if selectors == [] do
+        :all
+      else
+        selectors
+      end
+    end
+
+    defp remove_module_imports(%MapSet{} = current_imports, imported_module) do
+      current_imports
+      |> Enum.reject(&match?({^imported_module, _, _}, &1))
+      |> MapSet.new()
+    end
+
+    defp already_imported?(%MapSet{} = current_imports, imported_module) do
+      Enum.any?(current_imports, &match?({^imported_module, _, _}, &1))
+    end
+
+    defp function_and_arity_to_mfa(current_module, fa_list) when is_list(fa_list) do
+      for {function, arity} <- fa_list do
+        {current_module, function, arity}
+      end
+      |> MapSet.new()
+    end
+
+    defp mfas_for(current_module, type) do
+      if Code.ensure_loaded?(current_module) do
```

Comment:
This function seems to have a high cost. I tried simplifying it by removing half of the `Code.ensure_loaded?` calls, and it reduced the indexing time from around 9 seconds to 6 seconds.

https://github.com/lexical-lsp/lexical/compare/function-references...scottming:lexical:optimize-code-load?expand=1

Reply:
If we change Code.ensure_loaded? in my branch to Code.load?, the result might surprise you as the time could be reduced to 2 seconds.
```elixir
Time: 1.9 seconds
```

However, there are not as many modules that cannot be loaded by this piece of code:
https://github.com/scottming/lexical/commit/b5f5564d671c9d1a224e0461f364b894c997a6b9#diff-3745d0615d31f6563c8a33a36e1f58bea8539aa6e594fad72eb0d7813cae3fc0R9
 as imagined. Only these modules:
 
 ```elixir
 [(common 0.3.0) lib/lexical/ast/analysis/analyzer.ex:57: Lexical.Ast.Analysis.Analyzer.Import.apply_to_scope/3]
put_unimported_module(import_module) #=> MapSet.new([Bitwise, ExUnit.Assertions, ExUnit.CaptureIO, ExUnit.CaptureLog,
 Kernel.SpecialForms, Lexical.Test.Entry.Builder, Lexical.Test.Fixtures,
 Lexical.Test.Protocol.Fixtures.LspProtocol, Mix.Generator, MyImports,
 Phoenix.Component, Phoenix.Component.Declarative, Phoenix.Controller,
 Phoenix.Controller.Pipeline, Phoenix.HTML, Phoenix.HTML.Tag, Phoenix.LiveView,
 Phoenix.Logger, Phoenix.Socket, Plug.Builder, Plug.Conn, Statistex,
 Supervisor.Spec, TestHelper])
 ```

So, I think we need to find a way to pre-load these modules instead of doing it inside `push_scope`.

Reply:
I've pushed a change that caches the results of `ensure_loaded` and reduces the indexing time to 4.4 seconds.